### ISSUE TITLE
ODIC 'auth-server-url' and 'client-id' or alternatively 'public-key' should not be mandatory for case "Custom TenantConfigResolver"

### ISSUE THREAD

**Issue by michael-schnell** [quarkusio/quarkus#8851]
**Describe the bug**
For security-openid-connect extensions, the properties 'auth-server-url' and 'client-id' or alternatively 'public-key' are always mandatory, even if they are not required in case of using a TenantConfigResolver. 

The following exception is thrown on startup if they are not configured:
```
io.quarkus.runtime.configuration.ConfigurationException: Both 'auth-server-url' and 'client-id' or alterntively 'public-key' must be configured when the quarkus-oidc extension is enabled
	at io.quarkus.oidc.runtime.OidcRecorder.createTenantContext(OidcRecorder.java:96)
	at io.quarkus.oidc.runtime.OidcRecorder.setup(OidcRecorder.java:52)
	at io.quarkus.deployment.steps.OidcBuildStep$setup25.deploy_0(OidcBuildStep$setup25.zig:64)
	at io.quarkus.deployment.steps.OidcBuildStep$setup25.deploy(OidcBuildStep$setup25.zig:89)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:179)
	at io.quarkus.runtime.Application.start(Application.java:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:177)

```

**Expected behavior**
In case that a [TenantConfigResolver](https://github.com/quarkusio/quarkus/blob/master/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/TenantConfigResolver.java) was provided, the properties should **not** be mandatory, because they will most likely not be used. The configuration is returned by the resolver and that means there is no need for a static configuration in the application.properties

**Possible solution**
The [OidcRecorder](https://github.com/quarkusio/quarkus/blob/master/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRecorder.java) class should be enhanced to check wether a custom TenantConfigResolver is available and only throw the exception if this is not the case.

**Comment by sberyozkin** [comment]
Hi @michael-schnell 
I believe you refer to the default tenant (the regular) configuration in the `application.properties`.
The way to resolve it is to set `quarkus.oidc.tenant-enabled=false`.
We have it enabled by default because `TenantConfigResolver` may return `null` - which that it wants a fallback to the default configuration.
So can you please try `quarkus.oidc.tenant-enabled=false` ?
If it works then please do not close this issue - I'll follow with a minor doc update because it is not the 1st time the issue has been raised
Cheers

**Comment by michael-schnell** [comment]
Actually I'm using ODIC multitenancy, so I guess it has to be enabled, or not? But I don't need that default as the TenantConfigResolver I implemented always returns something useful (non-null). 

By the way:  I just implemented [multitenancy for Hibernate](https://github.com/quarkusio/quarkus/pull/8545). Hibernate does not allow a null value for a tenantId and so the interface there will never allow NULL as "default". I originally thought of aligning the two interfaces, but NULL seems to cause more problems as "default" than it solves.

**Comment by sberyozkin** [comment]
@michael-schnell This is not related to Hibernate.
In `quarkus-oidc` we need to be able to fallback if either `TenantResolver` or `TenantConfigResolver` can not resolve a tenant id from the current request, and hence in case of `TenantConfigResolver` - return an initialized `OidcTenantConfig`.  If in your case it is not going to happen then the default tenant can be disabled.
What happens when you add `quarkus.oidc.tenant-enabled=false` ?

**Comment by michael-schnell** [comment]
So to sum it up, the flag "quarkus.oidc.tenant-enabled" should have been better named "quarkus.oidc.default-tenant-enabled" !?

I will try it and give you feedback.

**Comment by sberyozkin** [comment]
>So to sum it up, the flag "quarkus.oidc.tenant-enabled" should have been better named "quarkus.oidc.default-tenant-enabled" !?

No. Because this property applies to any tenant

**Comment by sberyozkin** [comment]
@michael-schnell FYI, I've opened PR which will resolve this issue. Please reopen if needed. Thank you.

**Comment by michael-schnell** [comment]
Works fine. Tnx.

**Comment by michael-schnell** [comment]
Unfortunately I realized that it does not work...

Given I have:
`quarkus.oidc.tenant-enabled=false`

Then I get an exception:
```
io.quarkus.oidc.OIDCException: Tenant configuration context has not been resolved
	at io.quarkus.oidc.runtime.OidcAuthenticationMechanism.isWebApp(OidcAuthenticationMechanism.java:45)
	at io.quarkus.oidc.runtime.OidcAuthenticationMechanism.authenticate(OidcAuthenticationMechanism.java:32)
	at io.quarkus.oidc.runtime.OidcAuthenticationMechanism_ClientProxy.authenticate(OidcAuthenticationMechanism_ClientProxy.zig:279)
	at io.quarkus.vertx.http.runtime.security.HttpAuthenticator.attemptAuthentication(HttpAuthenticator.java:101)
	at io.quarkus.vertx.http.runtime.security.HttpAuthenticator_ClientProxy.attemptAuthentication(HttpAuthenticator_ClientProxy.zig:45)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$2.handle(HttpSecurityRecorder.java:93)
	at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$2.handle(HttpSecurityRecorder.java:45)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1034)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:131)
	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:130)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:54)
	at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:36)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$7.handle(VertxHttpRecorder.java:334)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$7.handle(VertxHttpRecorder.java:331)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$1.handle(VertxHttpRecorder.java:112)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$1.handle(VertxHttpRecorder.java:103)
	at io.vertx.core.http.impl.WebSocketRequestHandler.handle(WebSocketRequestHandler.java:50)
	at io.vertx.core.http.impl.WebSocketRequestHandler.handle(WebSocketRequestHandler.java:32)
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:136)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:369)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:232)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:173)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:101)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.vertx.core.http.impl.Http1xUpgradeToH2CHandler.channelRead(Http1xUpgradeToH2CHandler.java:109)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:321)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:295)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

The problem seems to be related to this method in "DefaultTenantConfigResolver":
```java
    private TenantConfigContext getTenantConfigFromConfigResolver(RoutingContext context, boolean create) {
        if (tenantConfigResolver.isResolvable()) {
            OidcTenantConfig tenantConfig;

            if (context.get(CURRENT_TENANT_CONFIG) != null) {
                tenantConfig = context.get(CURRENT_TENANT_CONFIG);
            } else {
                OidcTenantConfig newTenantConfig = this.tenantConfigResolver.get().resolve(context);
                if (newTenantConfig != null && !newTenantConfig.tenantEnabled) {
                    newTenantConfig = null;
                }
                tenantConfig = newTenantConfig;
                context.put(CURRENT_TENANT_CONFIG, tenantConfig);
            }

            if (tenantConfig != null) {
                String tenantId = tenantConfig.getTenantId()
                        .orElseThrow(() -> new IllegalStateException("You must provide a tenant id"));
                TenantConfigContext tenantContext = dynamicTenantsConfig.get(tenantId);

                if (tenantContext == null && create) {
                    synchronized (dynamicTenantsConfig) {
                        return dynamicTenantsConfig.computeIfAbsent(tenantId,
                                clientId -> tenantConfigBean.getTenantConfigContextFactory().apply(tenantConfig));
                    }
                }

                return tenantContext;
            }
        }

        return null;
    }
```

My tenant resolver returns correctly a "newTenantConfig" (it's non-null):
`OidcTenantConfig newTenantConfig = this.tenantConfigResolver.get().resolve(context);`

But the following lines do not work correctly:

```java
TenantConfigContext tenantContext = dynamicTenantsConfig.get(tenantId); // null
if (tenantContext == null && create) { // null && false
	synchronized (dynamicTenantsConfig) {
		return dynamicTenantsConfig.computeIfAbsent(tenantId,
				clientId -> tenantConfigBean.getTenantConfigContextFactory().apply(tenantConfig));
	}
}
return tenantContext; // null
```
So it seems the "dynamicTenantsConfig" map is not filled in case `quarkus.oidc.tenant-enabled=false`. 

Everything works fine in case this config is not set, but then I need to have the (unwanted) default config:
```
quarkus.oidc.auth-server-url=http://localhost:8180/auth/realms/never-used
quarkus.oidc.client-id=never-used

```

**Comment by sberyozkin** [comment]
@michael-schnell Oh sorry, I'll take care of it tomorrow, thanks

**Comment by michael-schnell** [comment]
Just another question: Why is the block with "dynamicTenantsConfig" synchronized? I've seen "dynamicTenantsConfig" is already a "ConcurrentHashMap", so it should be thread safe.

**Comment by sberyozkin** [comment]
@michael-schnell Originally there was a `HashMap` but even after ConcurrentHashMap was introduced we kept the sync block to avoid the redundant connection requests under a high load.

**Comment by sberyozkin** [comment]
@michael-schnell I've reproduced. There is a test which checks this flag is effective for the default configuration, but without the multi-tenancy, all the multi-tenancy tests have the enabled default configuration.

**Comment by sberyozkin** [comment]
@gsmet It probably won't make it to 1.4.2, I'm seeing some weird issues after my fix. Trying to isolate...

**Comment by michael-schnell** [comment]
No problem

**Comment by sberyozkin** [comment]
I think I've fixed it, the was a deadlock due to the changed blocking check condition...

**Comment by michael-schnell** [comment]
Cool

**Comment by sberyozkin** [comment]
Fixed for 1.5.0, thanks
